Fix #73: Adds Shields menu popup UI

kylehickinson opened the pull request:
Adds a shields menu UI rebuilt with modern API

## Pull Request Checklist

- [ ] My patch has gone through review and I have addressed review comments
- [x] My patch has a standard commit message that looks like `Fix #123: This fixes the shattered coffee cup!`

- [ ] I have updated the *Unit Tests* to cover new or changed functionality
- [ ] I have updated the *UI Tests* to cover new or changed functionality
- [x] I have marked the bug with `[needsuplift]`
- [x] I have made sure that localizable strings use `NSLocalizableString()`

## Screenshots

| Shields Off | Shields On |
| --- | --- |
| ![shieldsoff](https://user-images.githubusercontent.com/529104/43859140-d48d7d0a-9b1d-11e8-8f02-f9b4b10c25b6.png) | ![shieldson](https://user-images.githubusercontent.com/529104/43859154-e230e046-9b1d-11e8-8edf-0c7ab5a868da.png) |

## Notes for testing this patch

_None included_

Commit:
Fix #73: Adds Shields menu popup UI

Jacalz commented:
I must say that this does look really good after seeing the screenshots of it 👍

Commit:
Update popover to not pin to screen edges on iPhone regardless of size

- Also fixes constraints related to edge pinning and when the content size is larger than the screen width/height
- Adds 80 points to the bottom of the shields menu so it is easier to dismiss

Commit:
Hide the shields stats/controls when shields are off for a specific site

- Also hides the brave icon when no URL is loaded
- Fixes the height of a popover by taking the popover arrow into account

iccub commented:
I will do a code review later, ran this branch and found 3 things that could be improved.

1. Dismiss the popup on orientation change. Same goes for settings popup, I think we've talked about this for a bit. It looks fine on iPad but on iPhones the popup acts weird.
2. After disabling shields, the lion logo should be greyed out. Maybe this change fits better in some other adblock related task.
3. On iPad there's extra white space in the popup. Then after you disable shields, and reenable them, the white space is gone(meaning there should be an easy way to fix it). 

<img width="682" alt="zrzut ekranu 2018-08-09 o 15 27 26" src="https://user-images.githubusercontent.com/6950387/43901828-bf1aa8e6-9be8-11e8-9b56-b01acbfaef75.png">

Commit:
Automatically dismiss the popover when the orientation changes

Commit:
Fix issue where a popover may extend passed its required height

Due to the fact that the popover's content controller's view hasn't been computed yet, therefore we may not be able to compute its size for `preferredContentSize` correctly at the time of presentation

Commit:
Update the brave logo based on shields global "all off" state for a site

kylehickinson commented:
@iccub Got all 3 of those!

But as per https://github.com/brave/brave-ios/pull/135/commits/ae323c33593194bd18789e66f21e40fbc18116f9#diff-6d1fb0cf203fcd3c299d4ba3f07f21deR1431 we need to port proper cache drop, etc. with Shields. We don't currently use our own "BraveWebView" still.

Also waiting on #130 to add persistence to the brave shields site specific settings via [BraveShieldState.swift:L19](https://github.com/brave/brave-ios/blob/development/BraveShared/BraveShieldState.swift#L19)

jhreis commented on Client/Frontend/Browser/TabLocationView.swift:
```diff
@@ -53,12 +55,26 @@ class TabLocationView: UIView {
+    
+    /// Update the shields icon based on whether or not shields are enabled for this site
+    func refreshShieldsStatus() {
+        if let url = url, let domain = url.normalizedHost, let state = BraveShieldState.getStateForDomain(domain),
```
`url` is only used in the conditional here, so just do `domain = url?.normalizedHost` instead of unwrapping `url` itself.

kylehickinson replied:
Ah, good call

jhreis commented on Client/Frontend/Browser/TabLocationView.swift:
```diff
@@ -164,6 +180,17 @@ class TabLocationView: UIView {
         return pageOptionsButton
     }()
     
+    lazy var shieldsButton: ToolbarButton = {
```
`then`?

kylehickinson replied:
@iccub asked me the same thing, and I honestly thought we tabled `Then` in favour of consistency which is why I haven't been using it.

jhreis commented on Client/Frontend/Popover/PopoverController.swift:
```diff
@@ -316,17 +334,9 @@ extension PopoverController: BasicAnimationControllerDelegate {
-                containerView.rightAnchor.constraint(equalTo: viewController.view.safeAreaLayoutGuide.rightAnchor, constant: -outerMargins.right)
-            ])
+        containerView.snp.makeConstraints {
+            $0.left.greaterThanOrEqualTo(viewController.view.snp.left).offset(outerMargins.left).priority(UILayoutPriority.required.rawValue)
```
Isn't default `priority` == `required`?

kylehickinson replied:
...Yes...and yet I coulda swore when I was testing it was breaking constraints without this…

Investigation ensues lol

iccub commented on Client/Frontend/Popover/PopoverController.swift:
```diff
@@ -96,20 +105,23 @@ class PopoverController: UIViewController {
+            view.layoutIfNeeded()
+            
+            containerViewHeightConstraint = containerView.heightAnchor.constraint(equalToConstant: contentController.preferredContentSize.height + PopoverUX.arrowSize.height)
+            containerViewHeightConstraint?.priority = UILayoutPriority(rawValue: 850.0)
```
I would extract this 850 value to a constant.

jhreis replied:
Maybe even better, add it as an extension on the UILayoutPriority.

iccub commented on Client/Frontend/Popover/PopoverController.swift:
```diff
@@ -44,6 +44,9 @@ class PopoverController: UIViewController {
     var arrowDirectionBehavior: ArrowDirectionBehavior = .automatic
     
+    /// Whether or not to automatically dismiss the popup when the device orientation changes
+    var dismissesOnOrientationChanged = true
```
This is not used anymore right? I don't see any place where it changes its value. Or is it for some future-proofing for popups that won't be dismissed on orientation change.

kylehickinson replied:
Yeah I added it as a variable for future proofing rather than just force popover dismissal for every popover.

This may end up never being changed at which point we could probably remove it :)

iccub commented on Client/Frontend/Shields/ShieldsView.swift:
```diff
@@ -0,0 +1,286 @@
+    class View: UIView {
+        private let scrollView = UIScrollView()
+        
+        let stackView: UIStackView = {
```
Why not lazy vars for those views?

kylehickinson replied:
I prefer not to use `lazy var` for anything unless needed (i.e. using `self` within the declaration). Plus these are used on `init(frame:)` anyways so even if they were lazy they would be instantiated right on init anyways

iccub commented on Client/Frontend/Shields/ShieldsView.swift:
```diff
@@ -0,0 +1,286 @@
+            statViews.forEach {
+                addArrangedSubview($0)
+                if $0 !== statViews.last {
+                    setCustomSpacing(3.0, after: $0)
```
Didn't know about this method, great stuff.

iccub commented on Client/Frontend/Shields/ShieldsViewController.swift:
```diff
@@ -0,0 +1,164 @@
+    private func updateGlobalShieldState(_ on: Bool, animated: Bool = false) {
+        let updateBlock = {
+            self.shieldsView.shieldsContainerStackView.isHidden = !on
+            self.shieldsView.shieldsContainerStackView.alpha = on ? 1.0 : 0.0
```
Any reason why we set both `isHidden` and `alpha` values here?

kylehickinson replied:
Just animation purposes, looks nicer when its alpha fading at the same time rather than just being clipped while it does the `isHidden` animation.

jhreis replied:
Any reason we need to use `isHidden` at all?

iccub commented on Client/Frontend/Shields/ShieldsViewController.swift:
```diff
@@ -0,0 +1,164 @@
+        super.viewDidLayoutSubviews()
+        
+        shieldsView.stackView.layoutIfNeeded()
+        preferredContentSize = CGSize(width: 320.0, height: shieldsView.stackView.bounds.height)
```
I would extract this hardcoded width somewhere, we use it in two places.

iccub commented on Client/Frontend/Shields/ShieldsViewController.swift:
```diff
@@ -0,0 +1,164 @@
+    }
+    
+    override func loadView() {
+        view = View()
```
Why do we need to do this manually?

kylehickinson replied:
Because I am using a custom view (`ShieldsViewController.View`) rather than let `UIViewController` create the view.

iccub replied:
Oh, I just noticed this is a `View` not `UIView`..

kylehickinson replied:
Maybe I should rename it heh… Using nested classes lets me name it `View` and not worry about collisions but if its too confusing I will just rename it to `ShieldsView`

iccub replied:
It's harder to see on github, in XCode custom classes are highlighted in different color so it's fine for me.

iccub commented on Client/Frontend/Shields/ShieldsViewController.swift:
```diff
@@ -0,0 +1,164 @@
+        (.FpProtection, shieldsView.shieldsContainerStackView.fingerprintingControl, Preferences.Shields.fingerprintingProtection),
+    ]
+    
+    var shieldsView: View {
```
Can't we just use uiviewcontroller's `view` directly? Or is it done for better naming clarity

kylehickinson replied:
We could, the underlying memory of `view` would point to `shieldsView` however we'd require a cast each time, so this is just a convenience getter

iccub commented on Client/Frontend/Shields/ShieldsViewController.swift:
```diff
@@ -0,0 +1,164 @@
+    private func updateBraveShieldState(shield: BraveShieldState.Shield, on: Bool) {
+        guard let domain = url.normalizedHost else { return }
+        if shield == .AllOff {
+            // Technically we set "all off" when the switch is OFF, unlike all the others
```
Looks like we have to do a lot of exceptions to handle `.AllOff` state. Do you think it would be worth to remove it from shields state? Just wondering.

kylehickinson replied:
Yeah this is weird to be honest. I can't really remove it from the shields state because it does need to be saved & persisted along with other site-specific shield settings.

I could change it to be `AllOn` but it gets weird on the shield state side since we have to add things there returning true if its _not_ in the `state` dictionary.

All-around kind of odd, but I think I'd rather have the one-off checks be on the UI side than the data side 😅

Commit:
Move reused popover constants into declared variables

kylehickinson commented:
@jhreis Not sure why I can't reply to that last one but we need to use `isHidden` because thats what `UIStackView` uses to collapse/expand views, the alpha part is purely cosmetic

Commit:
Remove unused constraint priorities for edge-pinning popovers

+ Cleanup unused variables in an if-let

tmancey commented:
I have checked comments and am happy to merge. Joel asked me to merge on Friday so we can close down 0.2.1
